dll/win32/kernel32/client/handle.c

```diff
@@ -187,6 +221,145 @@ DuplicateHandle(IN HANDLE hSourceProcessHandle,
+		if (isSocket && GetCurrentProcessId() == GetProcessId(hSourceProcessHandle)) 
+        {
+			canDupe = TRUE; // this is the only case that is fairly certain to work
+		}
+		if (isSocket && hGetParentProcessId() == GetProcessId(hSourceProcessHandle)) { // handle is from parent process, it can be directly duplicated if inheritable
+			// check if handle has inheritance set
+			if (GetHandleInformation(hSourceHandle, &handleFlags)) { // get flags for handle
+				if (handleFlags & HANDLE_FLAG_INHERIT) { // need to check if it's winsock.. or not.. also if access is changing or inheritance is changing, thos would require dupe
+					if (dwOptions & DUPLICATE_SAME_ACCESS) {
+						DPRINT("DuplicateHandle: winsock inherited handle with same access, duplicating handle.\n");			
+					} else {
+						DPRINT("DuplicateHandle: winsock inherited handle with differing access, duplicating handle and ignoring access/dwOptions (PART IMPL FIXME).\n");										
+					}
+					canDupe = TRUE;
+				}
+			} else {
+				DPRINT1("DuplicateHandle: GetHandleInformation failed for parent owned handle(%lx) getlasterror is (%lx).\n", hSourceHandle, GetLastError());			
+			}
+		}
+		if(isSocket && canDupe && winsock_module ) 
+        { 
+			SOCKET targetSocket;
+			*lpTargetHandle = hSourceHandle; // if we fail we just return the handle we were given HACK TODO: Fix this
+			if (dupwinsock) 
+            { // if we didn't dynlink this gets skipped
+				_SEH2_TRY 
+				{			
+					int result = dupwinsock((SOCKET)hSourceHandle, GetCurrentProcessId(), &SharedSocketInfo); // was GetProcessId(ProcessHandle)
+					if (result != 0) {
+						DPRINT1("DuplicateHandle: IN Process call to WSADuplicateSocketW failed, returned:%x.\n", result);
+					} else {
+						// now call wsasocket to finish the process and create new socket
+						targetSocket = DynWSASocket(SharedSocketInfo.iAddressFamily, SharedSocketInfo.iSocketType, SharedSocketInfo.iProtocol, &SharedSocketInfo, 0, WSA_FLAG_OVERLAPPED); 
+						// 	not sure we want overlapped set, it's the default for winsock 1.x but winsock 2 stopped setting it by default	
+						if (targetSocket != INVALID_SOCKET) {
+							DPRINT("DuplicateHandle: winsock socket handle succesfully duplicated, returning handle(%lx).\n", targetSocket);
+							*lpTargetHandle = (HANDLE)targetSocket;							
+						} else DPRINT1("DuplicateHandle: WSASocketW on cross process call returned INVALID_SOCKET, returning input handle(%lx) (BADIMPL FIXME).\n", hSourceHandle);						
+					}
+				}			
+				_SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+				{
+					DPRINT1("DuplicateHandle: SOCKET DUPLICATION exception (%lx) for duplicating handle(%lx) returning source handle (BADIMPL FIXME).\n", _SEH2_GetExceptionCode(), hSourceHandle);			
+				}
+				_SEH2_END;		
+			}
+			DPRINT("DuplicateHandle: SOCKET DUPLICATION completed, returning true.\n");									
+			return TRUE;																
+		} else 
+        {
+			if (isSocket && winsock_module && !canDupe) 
```

Comment:
`else` is superfluous after `return`.
You could do 1 `if (isSocket)` instead of repeating that condition multiple times. Probably for `winsock_module` too.

Reply:
the else is outside the code block with the return.. I will look at it in context but it looks valid (but badly indented) at first blush.

The repeated if takes place in a for loop and the first if ensures that section of code in it executes only on the first pass of the loop, while the second if essentially skips that section of code if first section fails. Thinking about it again though it may be possible to get rid of the second if and just break out of the loop if the code inside the first if section fails.

Reply:
You are right about the if.. the first one for !winsock_module is definitely superfluous. The code is not in a loop here (it is in the other PR). I'll look it over more carefully, but the intent was to treat three cases in separate code blocks because one is very likely to need future revision (the one for !canDupe).

Reply:
Rebuilt the if structure. I saw what you meant about the superfluous parts after I reoriented things as you suggested. Please look at the revisions and see if they match what you expected now. Sorry, I have looked at it so many times I'm code blind.